### PR TITLE
Added feedstock for kubelogin

### PR DESCRIPTION
**Pull request by krisevans**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml"
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/master/recipes/example/meta.yaml#L57-L66) for an example)
- [x] Source is from official source
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged)
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Build number is 0
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details)
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there

### CODE REVIEW

**Comment by dhirschfeld** [comment]
```
'kubelogin' is not recognized as an internal or external command,
operable program or batch file
```
`conda-build` packages up artefacts under `%PREFIX%`. I think what's happening here is that `make` is putting the binary elsewhere so it's not being packaged.

I think we may need to ignore the build script and just call the `go` compiler directly.

We can use another `go` recipe as a template:
https://github.com/conda-forge/terraform-provider-azurerm-feedstock/blob/master/recipe/meta.yaml

It looks like CF sets some variables which can be used:
https://github.com/conda-forge/terraform-provider-azurerm-feedstock/blob/723dd977cac2779e7fadabb5d0c3d4e988c5b1b1/.ci_support/win_64_.yaml#L5-L16

...and something like the below should (hopefully) work:
```yaml
build:
  number: 0
  script:
    - go build -v -o {{ target_gobin }}{{ name }}{{ target_goexe }}
```

**Comment by dhirschfeld** [comment]
It looked like it worked *but* the version information is missing:
```
(%PREFIX%) %SRC_DIR%>kubelogin --version 
kubelogin version 
git hash: 
Go version: 
Build time:
```
...which, I think, is because we're not setting the [`LDFLAGS`](https://github.com/Azure/kubelogin/blob/1ee0da905a5b9ce9dceacff7e914b3c3c1e198ea/Makefile#L20-L22):
```
LDFLAGS    := -X main.version=$(VERSION) \
    -X main.goVersion=$(shell go version | cut -d " " -f 3) \
	-X main.buildTime=$(BUILD_TIME)

<snip>
go build -o $(BIN) -ldflags "$(LDFLAGS)"
```

**Comment by dhirschfeld** [comment]
The version can be pulled from the `PKG_VERSION` env var
```
%SRC_DIR%>set "PKG_VERSION=0.0.5" 
```

...but the others might be more difficult.
```
> conda list '^go$' --json | jq .[0].version
"1.13.15"
```

**Comment by dhirschfeld** [comment]
Looks good!
```
(%PREFIX%) %SRC_DIR%>kubelogin --version 
kubelogin version 
git hash: 0.0.5
Go version: go1.13.15
Build time: 2020-09-04T06:26:33Z
```

**Review by dhirschfeld — APPROVED** [review]
:shipit:

**Comment by dhirschfeld** [comment]
From [gitter](https://gitter.im/conda-forge/conda-forge.github.io?at=5f549c62ec534f584fe742be) it seems we may have to try and find *all* of the required licenses.
https://github.com/conda-forge/conda-forge.github.io/issues/1052

...at least for `go` it seems there's a tool to help with this:
https://github.com/google/go-licenses#complying-with-license-terms

**Comment by dhirschfeld** [comment]
Might be something to crib from the [`argo` PR](https://github.com/conda-forge/staged-recipes/pull/11799)
https://github.com/hadim/staged-recipes/blob/d6d00d1415856f8c0d45d8dc6605b83afc16edef/recipes/argo/build_library_licenses.sh

File: recipes/kubelogin/build_library_licenses.sh [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+export KUBELOGIN_PATH="$TEMP_DIR/kubelogin"
+
+go get -v github.com/google/go-licenses
+
+cd $TEMP_DIR
+git clone https://github.com/Azure/kubelogin.git
+cd kubelogin/
+
+rm -fr $LIBRARY_LICENCES_PATH
+$GOBIN/bin/go-licenses save $KUBELOGIN_PATH --save_path=$LIBRARY_LICENCES_PATH
```
Comment (dhirschfeld) [inline recipes/kubelogin/build_library_licenses.sh:16]:
```suggestion
$GOBIN/go-licenses save $KUBELOGIN_PATH --save_path=$LIBRARY_LICENCES_PATH
```
One too many bins!
```
<snip>placehold_p/bin/bin/go-licenses
```

**Comment by dhirschfeld** [comment]
*All* the licenses output by `go-licenses` tool are now packaged in the `recipe/library_licenses` folder:

image


ping @chrisburr - is this good to merge?

File: recipes/kubelogin/build_library_licenses.sh [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+export GOBIN=$(go env GOBIN)
+export LIBRARY_LICENCES_PATH="$RECIPE_DIR/library_licenses/"
+export TEMP_DIR=$(mktemp -d)
+export KUBELOGIN_PATH="$TEMP_DIR/kubelogin"
+
+go get -v github.com/google/go-licenses
+
+cd $TEMP_DIR
+git clone https://github.com/Azure/kubelogin.git
```
Comment (scopatz) [inline recipes/kubelogin/build_library_licenses.sh:12]: Please use the sources section in the meta.yaml to get the source code

File: recipes/kubelogin/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  url: https://github.com/Azure/{{ name }}/archive/v{{ version }}.tar.gz
+  sha256: 666557ddb0b75ca1ccd8c4f0b872429efa7f5b6739485b00326b52a1ebee0a24
+
+build:
+  number: 0
+
+requirements:
+  build:
+    - {{ compiler('go') }}
+    - posix  # [win]
```
Comment (scopatz) [inline recipes/kubelogin/meta.yaml:18]: This seems to have bash dependence as well

**Review by scopatz — CHANGES_REQUESTED** [review]
Use sources

File: recipes/kubelogin/build_library_licenses.sh [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+cd kubelogin/
+
+rm -fr $LIBRARY_LICENCES_PATH
+$GOBIN/go-licenses save $KUBELOGIN_PATH --save_path=$LIBRARY_LICENCES_PATH
+
+cd $RECIPE_DIR
+rm -fr $TEMP_DIR
```
Comment (dhirschfeld) [inline recipes/kubelogin/build_library_licenses.sh:19]:
```suggestion
rm -fr $LIBRARY_LICENCES_PATH
pushd %SRC_DIR%
$GOBIN/go-licenses save $KUBELOGIN_PATH --save_path=$LIBRARY_LICENCES_PATH
popd
```

**Comment by dhirschfeld** [comment]
@scopatz - is there a better place to save the licenses? The *actual* license is in a `info/license` folder alongside the `info/recipe` folder but I don't know how we could access that folder from the build to e.g. add `thirdparty` subfolder?

**Comment by dhirschfeld** [comment]
The `posix` package includes `m2-base` which includes `m2-bash`:
```
posix 1.0.0 2
-------------
file name   : posix-1.0.0-2.tar.bz2
name        : posix
version     : 1.0.0
build       : 2
build number: 2
size        : 2 KB
subdir      : win-64
url         : https://repo.anaconda.com/pkgs/msys2/win-64/posix-1.0.0-2.tar.bz2
md5         : 83ce222b94ccaf79838a6ffa5a1cd6bc
dependencies:
  - m2-base
  - m2-diffutils
  - m2-make
  - m2-tar
  - m2-unzip
  - m2-zip
  - msys2-conda-epoch >=20160418
```

File: recipes/kubelogin/build_library_licenses.sh [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+go get -v github.com/google/go-licenses
+
+cd $TEMP_DIR
+git clone https://github.com/Azure/kubelogin.git
+cd kubelogin/
+
+rm -fr $LIBRARY_LICENCES_PATH
+pushd $SRC_DIR
+$GOBIN/go-licenses save $KUBELOGIN_PATH --save_path=$LIBRARY_LICENCES_PATH
+popd
```
Comment (dhirschfeld) [inline recipes/kubelogin/build_library_licenses.sh:18]:
```suggestion
export THIRDPARTY_LICENCES_PATH="$RECIPE_DIR/thirdparty_licenses/"

rm -fr $THIRDPARTY_LICENCES_PATH
go get -v github.com/google/go-licenses
$GOBIN/go-licenses save $SRC_DIR --save_path=$THIRDPARTY_LICENCES_PATH
```

**Review by dhirschfeld — APPROVED** [review]
LGTM

**Comment by scopatz** [comment]
> but I don't know how we could access that folder from the build to e.g. add thirdparty subfolder?

Yeah, you'd have to install it into share or something

File: recipes/kubelogin/meta.yaml (hunk 3 above)
Comment (scopatz) [reply recipes/kubelogin/meta.yaml:18]: This still needs a bash dep

Comment (scopatz) [reply recipes/kubelogin/meta.yaml:18]: At least on posix

**Comment by chrisburr** [comment]
> is there a better place to save the licenses? The _actual_ license is in a `info/license` folder alongside the `info/recipe` folder but I don't know how we could access that folder from the build to e.g. add `thirdparty` subfolder?

You could concatenate them together and the specify `license_file: ALL_LICENSES` in the `meta.yaml`.

```bash
cp "${SRC_DIR}/LICENSE" ALL_LICENSES
for fn in thirdparty_licenses/*; do
    echo "${fn}
    cat "${fn}"
done >> ALL_LICENSES
```

Alternatively, the `license_file:` field does accept a list so the files could be included individually, but I don't think it supports globbing to more easily include them all:

```yaml
  license_file:
    - LICENSE
    - thirdparty_licenses/*
```

**Comment by dhirschfeld** [comment]
Since the licenses are defined in named subfolders you would have to recursively concatenate all the licenses whilst also attributing the source (defined by the folder hierarchy).

Given that, I'm inclined to leave it as-is if that's not a problem? Whilst it might not be ideal, I believe it meets the license *requirements* and I think the right way to solve this would be adding globbing as a feature.

**Comment by dhirschfeld** [comment]
If someone is about who could kick the CI I think this PR will pass now that `go-licenses` is available.

**Comment by chrisburr** [comment]
@conda-forge-admin, please restart ci

> Given that, I'm inclined to leave it as-is if that's not a problem?

Are all of the files finding they way into the package? It looks like they're just being copied to the recipe directory which is then deleted when the CI job finishes though I might be missing something.

**Comment by dhirschfeld** [comment]
> *It looks like they're just being copied to the recipe directory*

`conda` packages everything in the recipe directory into the `info/recipe` folder in the tarball so all the licenses make it into the `bz2` package:

image

☝️ from the `go-licenses` recipe which uses itself to package its own licenses!

I remember, in the early days of `conda`, putting my recipe in the root of my project source and eventually finding the entire source tree in the final package! 😆

**Comment by dhirschfeld** [comment]
@conda-forge-admin, please restart ci

**Comment by dhirschfeld** [comment]
When using the packaged `go-licenses` binary it fails with the below error:
```
login
+ go-licenses save C:/Miniconda/conda-bld/kubelogin_1600256405799/work '--save_path=D:\a\1\s\recipes\kubelogin/thirdparty_licenses/'
Error: cannot register licenses from archive: open D:\bld\go-licenses_1600240623892\work\gopath\pkg\mod\github.com\google\licenseclassifier@v0.0.0-20190926221455-842c0d70d702\licenses\licenses.db: The system cannot find the path specified.
Usage:
  licenses save <package> [flags]

Flags:
      --force              Delete the destination directory if it already exists.
  -h, --help               help for save
      --save_path string   Directory into which files should be saved that are required by license terms

Global Flags:
      --alsologtostderr                  log to standard error as well as files
      --confidence_threshold float       Minimum confidence required in order to positively identify a license. (default 0.9)
      --log_backtrace_at traceLocation   when logging hits line file:N, emit a stack trace (default :0)
      --log_dir string                   If non-empty, write log files in this directory
      --logtostderr                      log to standard error instead of files
      --stderrthreshold severity         logs at or above this threshold go to stderr (default 2)
  -v, --v Level                          log level for V logs
      --vmodule moduleSpec               comma-separated list of pattern=N settings for file-filtered logging

F0916 11:43:49.397416    4972 main.go:43] cannot register licenses from archive: open D:\bld\go-licenses_1600240623892\work\gopath\pkg\mod\github.com\google\licenseclassifier@v0.0.0-20190926221455-842c0d70d702\licenses\licenses.db: The system cannot find the path specified.
```

I've not been able to find out why or fix the problem :/

**Comment by dhirschfeld** [comment]
I reverted to the last good commit over in #12674. Could that one please be merged and this one closed?

It might take a `go` expert to figure out why the packaged binary doesn't work.

**Pull request closed without merging**